File: androidApp/src/main/java/com/papslabs/omdb_kmp/android/ui/authentication/SignUpViewModel.kt
package com.papslabs.omdb_kmp.android.ui.authentication

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.papslabs.omdb_kmp.android.domain.state.SignUpScreenState
import com.papslabs.omdb_kmp.android.domain.state.TextFieldState
import com.papslabs.omdb_kmp.domain.model.Resource
import com.papslabs.omdb_kmp.domain.use_case.SignUpUseCase
import com.papslabs.omdb_kmp.util.extensions.isValidEmail
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.UUID

class SignUpViewModel(
    private val signUpUseCase: SignUpUseCase
): ViewModel() {

    // Global
    private val TAG = SignUpViewModel::class.java.simpleName
    var state by mutableStateOf(SignUpScreenState())
        private set
    var isLoading by mutableStateOf(false)
        private set
    var hasSignUpSuccessfully by mutableStateOf(false)
        private set
    var error by mutableStateOf<String?>(null)
        private set

    fun updateFirstName(firstName: String) {
        state = state.copy(
            firstName = TextFieldState(
                text = firstName.filter { it.isLetter() }
            )
        )
    }

    fun updateLastName(lastName: String) {
        state = state.copy(
            lastName = TextFieldState(
                text = lastName.filter { it.isLetter() }
            )
        )
    }

    fun updateEmail(email: String) {
        state = state.copy(
            email = TextFieldState(
                text = email.trim()
            )
        )
    }

    fun signUp() {
        if (isInvalid()) return
        signUpUseCase(
            id = UUID.randomUUID().toString(),
            firstName = state.firstName.text,
            lastName = state.lastName.text,
            email = state.email.text
        )
            .onEach {
                when (it) {
                    is Resource.Loading -> isLoading = true
                    is Resource.Success -> {
                        isLoading = false
                        hasSignUpSuccessfully = true
                    }
                    is Resource.Error -> {
                        error = it.msg
                        isLoading = false
                    }
                }
            }
            .launchIn(viewModelScope)
    }

    private fun isInvalid(): Boolean {
        val firstName = state.firstName.text.trim()
        val firstNameError = if (firstName.isEmpty()) "" else null

        val lastName = state.lastName.text.trim()
        val lastNameError = if (lastName.isEmpty()) "" else null

        val email = state.email.text.trim()
        val emailError = if (email.isEmpty() || !email.isValidEmail()) "" else null

        state = state.copy(
            firstName = state.firstName.copy(error = firstNameError),
            lastName = state.lastName.copy(error = lastNameError),
            email = state.email.copy(error = emailError)
        )
        return firstNameError != null
                || lastNameError != null
                || emailError != null
    }
}

File: iosApp/iosApp/ui/auth/SignUpViewModel.swift
//
//  SignUpViewModel.swift
//  iosApp
//
//  Created by Pranit Rane on 01/09/24.
//  Copyright © 2024 orgName. All rights reserved.
//
import Foundation
import shared

extension SignUpScreen {
    @MainActor class SignUpViewModel: ObservableObject {
        
        // Global
        private let TAG = "SignUpViewModel"
        private let signUpUseCase: SignUpUseCase  = UseCaseHelper().signUpUseCase
        @Published private(set) var isLoading = false
        @Published private(set) var hasSignUpSuccessfully = false
        @Published private(set) var error: String? = nil
        @Published var firstName = ""
        @Published var lastName = ""
        @Published var email = ""
        
        func signUp() {
            if isInvalid() {
                return
            }
            signUpUseCase.invoke(
                id: NSUUID().uuidString,
                firstName: firstName.filterOutNonLetters(),
                lastName: lastName.filterOutNonLetters(),
                email: email.trimmingCharacters(in: .whitespacesAndNewlines)
            )
            .onEach(
                onLoading: {
                    self.isLoading = true
                },
                onSuccess: { user in
                    self.isLoading = false
                    self.hasSignUpSuccessfully = true
                },
                onError: { error in
                    self.error = error
                    self.isLoading = false
                })
        }
        
        private func isInvalid() -> Bool {
            let firstName = firstName.filterOutNonLetters()
            let firstNameError: String? = if firstName.isEmpty {
                ""
            } else {
                nil
            }
            
            let lastName = lastName.filterOutNonLetters()
            let lastNameError: String? = if lastName.isEmpty {
                ""
            } else {
                nil
            }
            
            let email = email.trimmingCharacters(in: .whitespacesAndNewlines)
            let emailError: String? = if email.isEmpty || !email.isValidEmail() {
                ""
            } else {
                nil
            }
            return firstNameError != nil
            || lastNameError != nil
            || emailError != nil
        }
    }
}
